modules/util/path.go

```diff
@@ -248,3 +240,28 @@ func IsReadmeFileExtension(name string, ext ...string) (int, bool) {
 
 	return 0, false
 }
+
+// SafeJoinPath is like path.Join, but it will prevent directory escaping.
+func SafeJoinPath(elem ...string) string {
+	elems := make([]string, len(elem))
+	for i, v := range elem {
+		elems[i] = path.Clean("/" + v)
+	}
+	if len(elem) > 0 && !strings.HasPrefix(elem[0], "/") {
+		return path.Join(elems...)[1:]
+	}
+	return path.Join(elems...)
+}
+
+// SafeJoinFilepath is like filepath.Join, but it will prevent directory escaping.
+func SafeJoinFilepath(elem ...string) string {
+	separator := string(filepath.Separator)
+	elems := make([]string, len(elem))
+	for i, v := range elem {
+		elems[i] = filepath.Clean(separator + v)
+	}
+	if len(elem) > 0 && !strings.HasPrefix(elem[0], separator) {
+		return filepath.Join(elems...)[1:]
+	}
```

Comment:
For windows: `C:\\foo` => `:\\foo` ?

I do not think Gitea has Windows servers to run tests .... so the test code for Windows never runs?

Reply:
My bad, I thought the CI would run tests on Windows ...

I will fix it and run the tests on Windows to ensure it works.